add scan function

arlor opened the pull request.

Commit:
add scan function

thautwarm commented:
I appreciate a lot for your pull request!  
However, this scan generator could be replaced by `Map`.
```python
def scan(f, start_elem):
      def call(now):
           nonlocal start_elem
           start_elem = f(now, start_elem)
           return start_elem
      return call

# `sequence` is not a stream, maybe a list, tuple, or dict and so on.
Flow(sequence).Map(scan(f, start_elem)) == Flow(sequence).Scan(f, start_elem)
```
I'd think for a while to decide whether to merge it.

Athedorer commented:
@thautwarm Thanks for your response! I use Scala in my daily work. Scan is one of the methods in the Scala standard library. I've  used it several times, it's convenient. So I guess someone might need this, too. Map is ok, but it's not clear, at least I didn't think of this. :joy:  I really appreciate Linq.py very much. Thanks for the work you have done!

thautwarm committed:
for immutable

do not change the existed instance of `ScanGenerator`.
Doing this costs nothing.

thautwarm committed:
Merge branch 'master' into master

thautwarm committed:
Merge branch 'master' into master

thautwarm commented:
@Athedorer add some tests to `test_manual.py`, please :)
